File: lib/imagewriter/imgimagewriter.cc
#include "lib/core/globals.h"
#include "lib/config/config.h"
#include "lib/config/flags.h"
#include "lib/data/sector.h"
#include "lib/imagewriter/imagewriter.h"
#include "lib/data/image.h"
#include "lib/config/proto.h"
#include "lib/config/config.pb.h"
#include "lib/data/layout.h"
#include "lib/config/layout.pb.h"
#include "lib/core/logger.h"
#include <algorithm>
#include <iostream>
#include <fstream>

class ImgImageWriter : public ImageWriter
{
public:
    ImgImageWriter(const ImageWriterProto& config): ImageWriter(config) {}

    void writeImage(const Image& image) override
    {
        const Geometry geometry = image.getGeometry();

        auto& layout = globalConfig()->layout();
        int tracks = layout.has_tracks() ? layout.tracks() : geometry.numTracks;
        int sides = layout.has_sides() ? layout.sides() : geometry.numSides;

        std::ofstream outputFile(
            _config.filename(), std::ios::out | std::ios::binary);
        if (!outputFile.is_open())
            error("cannot open output file");

        for (const auto& p : Layout::getTrackOrdering(tracks, sides))
        {
            int track = p.first;
            int side = p.second;

            auto trackLayout = Layout::getLayoutOfTrack(track, side);
            for (int sectorId : trackLayout->naturalSectorOrder)
            {
                const auto& sector = image.get(track, side, sectorId);
                if (sector)
                    sector->data.slice(0, trackLayout->sectorSize)
                        .writeTo(outputFile);
                else
                    outputFile.seekp(trackLayout->sectorSize, std::ios::cur);
            }
        }

        log("IMG: wrote {} tracks, {} sides, {} kB total to {}",
            tracks,
            sides,
            outputFile.tellp() / 1024,
            _config.filename());
    }
};

std::unique_ptr<ImageWriter> ImageWriter::createImgImageWriter(
    const ImageWriterProto& config)
{
    return std::unique_ptr<ImageWriter>(new ImgImageWriter(config));
}

File: java/com/cowlark/fluxengine/imagewriter/ImgImageWriter.java
package com.cowlark.fluxengine.imagewriter;

import com.cowlark.fluxengine.config.ConfigProto;
import com.cowlark.fluxengine.core.ByteWriter;
import com.cowlark.fluxengine.core.Bytes;
import com.cowlark.fluxengine.data.CylinderHead;
import com.cowlark.fluxengine.data.DiskLayout;
import com.cowlark.fluxengine.data.Geometry;
import com.cowlark.fluxengine.data.Image;
import com.cowlark.fluxengine.data.LogicalTrackLayout;
import com.cowlark.fluxengine.data.Sector;

/**
 * Writes a raw (non-interleaved) sector image, ported from
 * lib/imagewriter/imgimagewriter.cc.
 */
public class ImgImageWriter extends ImageWriter
{
    private final ConfigProto config;

    /* The img writer needs the full config to determine the layout; created
     * via ImageWriter.create(ConfigProto). */
    public ImgImageWriter(ImageWriterProto writerConfig, ConfigProto config)
    {
        super(writerConfig);
        this.config = config;
    }

    @Override
    public void writeImage(Image image)
    {
        Geometry geometry = image.getGeometry();

        int tracks = config.getLayout().hasTracks() ?
                config.getLayout().getTracks() :
                geometry.numCylinders;
        int sides =
                config.getLayout().hasSides() ? config.getLayout().getSides() : geometry.numHeads;

        DiskLayout diskLayout = new DiskLayout(config);
        boolean inFilesystemOrder = getWriterConfig().getImg().getFilesystemSectorOrder();

        Bytes output = new Bytes();
        ByteWriter bw = output.writer();

        Iterable<CylinderHead> locations = inFilesystemOrder ?
                diskLayout.logicalLocationsInFilesystemOrder :
                diskLayout.logicalLocations;
        for (CylinderHead logicalLocation : locations)
        {
            LogicalTrackLayout ltl = diskLayout.layoutByLogicalLocation.get(logicalLocation);

            Iterable<Integer> sectorOrder =
                    inFilesystemOrder ? ltl.filesystemSectorOrder : ltl.naturalSectorOrder;
            for (int sectorId : sectorOrder)
            {
                Sector sector =
                        image.get(logicalLocation.cylinder(), logicalLocation.head(), sectorId);
                if (sector != null)
                    bw.write(sector.data.slice(0, ltl.sectorSize));
                else
                    bw.pad(ltl.sectorSize);
            }
        }

        output.writeToFile(getWriterConfig().getFilename());

        System.out.printf(
                "IMG: wrote %d tracks, %d sides, %d kB total to %s%n",
                tracks,
                sides,
                output.size() / 1024,
                getWriterConfig().getFilename());
    }
}
